cli/command/telemetry.go

```diff
@@ -0,0 +1,214 @@
+}
+
+func (cli *DockerCli) TracerProvider(ctx context.Context, opts ...sdktrace.TracerProviderOption) TracerProvider {
+	exp, err := dockerSpanExporter(ctx, cli)
+	if err != nil {
+		otel.Handle(err)
+	}
+	defaultOpts := []sdktrace.TracerProviderOption{
+		sdktrace.WithResource(cli.Resource()),
+		sdktrace.WithBatcher(exp),
+	}
+	return sdktrace.NewTracerProvider(append(defaultOpts, opts...)...)
+}
+
+func (cli *DockerCli) MeterProvider(ctx context.Context, opts ...sdkmetric.Option) MeterProvider {
+	exp, err := dockerMetricExporter(ctx, cli)
+	if err != nil {
+		otel.Handle(err)
+	}
+	defaultOpts := []sdkmetric.Option{
+		sdkmetric.WithResource(cli.Resource()),
+		sdkmetric.WithReader(newCLIReader(exp)),
+	}
+	return sdkmetric.NewMeterProvider(append(defaultOpts, opts...)...)
+}
+
+// WithResourceOptions configures additional options for the default resource. The default
+// resource will continue to include its default options.
+func WithResourceOptions(opts ...resource.Option) CLIOption {
+	return func(cli *DockerCli) error {
+		cli.res.AppendOptions(opts...)
+		return nil
+	}
+}
+
+// WithResource overwrites the default resource and prevents its creation.
+func WithResource(res *resource.Resource) CLIOption {
+	return func(cli *DockerCli) error {
+		cli.res.Set(res)
+		return nil
+	}
+}
+
+type telemetryResource struct {
+	res  *resource.Resource
+	opts []resource.Option
+	once sync.Once
+}
+
+func (r *telemetryResource) Set(res *resource.Resource) {
+	r.res = res
+}
+
+func (r *telemetryResource) Get() *resource.Resource {
+	r.once.Do(r.init)
+	return r.res
+}
+
+func (r *telemetryResource) init() {
+	if r.res != nil {
+		r.opts = nil
+		return
+	}
+
+	opts := r.defaultOptions()
+	if len(r.opts) > 0 {
+		opts = append(opts, r.opts...)
+	}
+
+	res, err := resource.New(context.Background(), opts...)
+	if err != nil {
+		otel.Handle(err)
+	}
+	r.res = res
+
+	// Clear the resource options since they'll never be used again and to allow
+	// the garbage collector to retrieve that memory.
+	r.opts = nil
+}
+
+func (r *telemetryResource) defaultOptions() []resource.Option {
+	return []resource.Option{
+		resource.WithDetectors(serviceNameDetector{}),
+		resource.WithAttributes(
+			// Use a unique instance id so OTEL knows that each invocation
+			// of the CLI is its own instance. Without this, downstream
+			// OTEL processors may think the same process is restarting
+			// continuously.
+			semconv.ServiceInstanceID(uuid.Generate().String()),
+		),
+		resource.WithFromEnv(),
+		resource.WithTelemetrySDK(),
+	}
+}
+
+func (r *telemetryResource) AppendOptions(opts ...resource.Option) {
+	if r.res != nil {
+		return
+	}
+	r.opts = append(r.opts, opts...)
+}
+
+type serviceNameDetector struct{}
+
+func (serviceNameDetector) Detect(ctx context.Context) (*resource.Resource, error) {
+	return resource.StringDetector(
+		semconv.SchemaURL,
+		semconv.ServiceNameKey,
+		func() (string, error) {
+			return filepath.Base(os.Args[0]), nil
```

Comment:
Is the idea here to discern between `docker`/main CLI and plugin binaries?

Reply:
A bit. The service name is determined either through `OTEL_SERVICE_NAME` as an environment variable or through the default logic of `unknown_service:${args[0]}` as the name. This changes the default name to be just `args[0]` which usually maps to the name of the binary.

I'm still trying to determine what the service name is most useful for, but I've found it helpful for visual displays. I don't think it's useful for logic like "did this metric come from buildx?" because the user can overwrite the service name by using `OTEL_SERVICE_NAME`. At the moment, I've been using the instrumentation scope which maps to the tracer or meter name. I'm not sure if that will continue to work into the future.

This code specifically was copied from buildkit.